### ISSUE TITLE
Use the newly fixed debugging information

### ISSUE THREAD

**Issue by daxpedda** [rustwasm/wasm-bindgen#3698]
Just dropping some random thoughts before I forget them.

- Using it in `wasm-bindgen-test` instead of `console_error_panic_hook` (don't replace it entirely because we still need the panic message).
- Add a new documentation page noting how to actually deploy it.
  - Generate sourcemaps, e.g. using [`wasm2map`](https://crates.io/crates/wasm2map).
  - Strip DWARF in case of sourcemaps.
  - Notes on browser support and how to serve source files if not bundled.
- Maybe think about adding an option to generate source maps or split debug info directly in `wasm-bindgen`, so I would argue this would be more the domain of `wasm-pack` and similar.

**Comment by hamza1311** [comment]
A page in the documentation about this information would be very helpful

**Comment by xxshady** [comment]
for some reason wasm2map generates empty sourcemap file when i call it on .wasm generated by `wasm-pack`
`{"version":3,"names":[],"sources":[""],"sourcesContent":[],"mappings":""}`

commands i use:
`wasm-pack build --dev --target no-modules`
`cargo wasm2map --bundle-sources -p rust_wasm_bg.wasm -b http://127.0.0.1:8080`

it generates correct sourcemap with .wasm from target/wasm32-unknown-unknown/debug though

**Comment by daxpedda** [comment]
You have to use the [`--keep-debug` CLI option](https://rustwasm.github.io/docs/wasm-bindgen/reference/cli.html#--keep-debug) to not discard the DWARF information when post-processing with `wasm-bindgen`.

How to do that with `wasm-pack` is something you might want to ask there.

**Comment by xxshady** [comment]
yeah, thank you, with `--keep-debug` wasm2map generates correct sourcemap, i did not found how to enable it in wasm-pack, but using wasm-bindgen CLI directly it worked:
1. `cargo build --target wasm32-unknown-unknown`
2. `wasm-bindgen --target <your target> --out-dir pkg --keep-debug target/wasm32-unknown-unknown/debug/<crate name>.wasm`
(if you don't have wasm-bindgen CLI installed: `cargo install wasm-bindgen-cli`)
3. `cargo wasm2map pkg/<crate name>_bg.wasm`
(you probably also need to provide other flags if your target is browser or nodejs, see wasm2map [example](https://github.com/mtolmacs/wasm2map/blob/55978542eddc08cd43895999fb3ad5586020d837/example/run.bat#L7))
4. sourcemap should be available in `pkg/<crate name>_bg.wasm.map`

> note: i only tested it with `wasm-bindgen --target no-modules`, not sure if others will work the same

also, if you are building something outside of browser or nodejs (like me) where sourcemaps are not supported out of the box, here is how you can use sourcemap manually to get original source code line and column:

1. `cargo wasm2map pkg/<crate name>.wasm --bundle-sources --patch --base-url blabla`
(`--base-url blabla` because it doesnt matter anyway, but required by `--bundle-sources`)

2. for example we have this frame:
`at rust_wasm.wasm.rust_wasm::script_events::test_script_events::____test_panic4____::he9b091250693a65d (wasm://wasm/rust_wasm.wasm-01976b8a:wasm-function[2428]:0xbf5ed)`
`0xbf5ed` is the column, the line is always 1 (because .wasm is binary as i understand)
3. now you can pass line and colum your sourcemap resolver (anything that can resolve sourcemaps v3)
4. i'm gonna use [`source-map-cli`](https://www.npmjs.com/package/source-map-cli) for it
5. `source-map-cli` doesn't support hex numbers, so converting column to decimal first: 783853
6. `npx source-map-cli resolve pkg/rust_wasm_bg.wasm.map 1 783853`
```
Maps to .../src/script_events.rs:300:4 

    panic!("test panic");
    ^
```
> note: `cargo wasm2map pkg/<crate name>_bg.wasm` is already enough to get sourcemap but `source-map-cli` requires bundled source code

**Comment by andrewbaxter** [comment]
I didn't quite get this working.  IIUC the current required steps are:

1. Build with debug info (debug mode, or debug=true in profile.release)
2. Run wasm-bindgen with --keep-debug
3. Do wasm2map.  I did it via the rust library, and for `x.wasm` wrote to `x.wasm.map` and patched with url `"x.wasm.map"` (no absolute url).

I got the same wasm bytecode backtrace, no rust line numbers.

Clicking on the lines _did_ attempt to pull up the source map, but the change to wasm2map to include the source files hasn't been released yet.

I did 1-3 above plus `console-error-panic-hook` and no improvement.

**Comment by daxpedda** [comment]
We were having a lengthy discussion in #4635 about if we even want to go with source maps.

Cc @Spxg, feels rude to just close your issue like that, sorry!
